numpy/core/src/umath/ufunc_object.c

```diff
@@ -1731,9 +1746,145 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GET_ITEM(axes, i);
+        item_is_tuple = PyTuple_Check(item);
+        if (item_is_tuple) {
+            num_axes = PyTuple_Size(item);
+        }
+        else {
+            /* try as integer */
+            requested_axis = (int)PyInt_AsSsize_t(item);
+            if (error_converting(requested_axis)) {
+                PyErr_Clear();
+                /* to give descriptive error message below */
+                num_axes = -1;
+            }
+            else {
+                num_axes = 1;
+            }
+        }
+        if (num_axes != core_ndim) {
+            if (core_ndim == 1) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with a "
+                             "single element, an integer, or None", i);
+            }
+            else {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with %d "
+                             "elements, or None", i, core_ndim);
+            }
+            return -1;
+        }
+        /*
+         * Now create the remap, starting with the core dimensions, and
+         * then adding the remaining axes that are to be iterated over.
+         */
+        for (j = 0; j < op_ndim; j++) {
+            have_seen_axis[j] = 0;
+        }
+        for (j = 0; j < core_ndim; j++) {
+            if (item_is_tuple) {
+                requested_axis = (int)PyInt_AsSsize_t(
+                    PyTuple_GET_ITEM(item, j));
+                if (error_converting(requested_axis)) {
+                    return -1;
+                }
+            }
+            if (check_and_adjust_axis(&requested_axis, op_ndim) < 0) {
+                return -1;
+            }
+            if (have_seen_axis[requested_axis]) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d has value %d repeated",
+                             i, requested_axis);
+                return -1;
+            }
+            have_seen_axis[requested_axis] = 1;
+            remap_axis[i][op_ndim - core_ndim + j] = requested_axis;
+        }
+        iter_axis = 0;
+        for (j = 0; j < op_ndim - core_ndim; j++) {
+            while (have_seen_axis[iter_axis]) {
+                iter_axis++;
+            }
+            remap_axis[i][j] = iter_axis++;
+        }
+        /*
+         * Check whether we are actually remapping anything. Here,
+         * iter_axis can only equal j if all broadcast axes were the same.
+         */
+        if (iter_axis == j) {
+            while (j < op_ndim && remap_axis[i][j] == j) {
+                j++;
+            }
+        }
+        do_remap[i] = (j < op_ndim);
```

Comment:
Are you sure that doing this check is more efficient than just doing a no-op remap?